shiny/_template_utils.py

```diff
@@ -0,0 +1,150 @@
+        "default": lambda: appTemplateQuestions(template),
+    }
+
+    # Call the appropriate function based on the user's choice
+    flow_dispatch.get(template, flow_dispatch["default"])()
+
+
+def appTemplateQuestions(template: str):
+    appdir = questionary.path(
+        "Enter destination directory:",
+        default="./",
+        only_directories=True,
+    ).ask()
+
+    app_dir = copyTemplateFiles(appdir, template)
+    print(f"Created Shiny app at {app_dir}")
+
+
+def jsComponentQuestions():
+    """
+    Hand question branch for the custom js templates. This should handle the entire rest
+    of the question flow and is responsible for placing files etc. Currently it repeats
+    a lot of logic from the default flow but as the custom templates get more
+    complicated the logic will diverge
+    """
+
+    component_type = questionary.select(
+        "What kind of component do you want to build?:",
+        choices=choicesArray(
+            [
+                ("Input component", "js-input"),
+                ("Output component", "js-output"),
+                ("React component", "js-react"),
+                ("Back", "back"),
+            ]
+        ),
+    ).ask()
+
+    if component_type == "back":
+        template_query()
+        return
+
+    # As what the user wants the name of their component to be
+    component_name = questionary.text(
+        "What do you want to name your component?",
+        instruction="Name must be dash-delimited and all lowercase. E.g. 'my-component-name'",
+        validate=isValidName,
+    ).ask()
+
+    appdir = questionary.path(
+        "Enter destination directory:",
+        default=f"./{component_name}",
+        only_directories=True,
+    ).ask()
+
+    app_dir = copyTemplateFiles(appdir, component_type)
+
+    # Print messsage saying we're building the component
+    print(f"Setting up {component_name} component package...")
+    updateComponentNameInTemplate(app_dir, component_name)
+
+    shouldInstallDeps = questionary.confirm(
+        "Do you want to install js dependencies now?"
+    ).ask()
+
+    if shouldInstallDeps:
+        install_js_dependencies(app_dir)
+    else:
+        print("Skipping installing NPM deps. Run `npm install` to install them later.")
+
+    print(f"Successfully created {component_name} component package!")
+    print("Next steps:")
+    print(f"- Run `cd {app_dir}` to change into the new directory")
+    if not shouldInstallDeps:
+        print("- Run `npm install` to install dependencies")
+    print("- Run `npm run build` to build the component")
+    print("- Install package locally with `pip install -e .`")
+    print("- Open and run the example app in the `example-app` directory")
+
+
+def choicesArray(choices: list[tuple[str, str]]):
+    """
+    Convert tuple of key-value pairs for Questionary questions to a list of Choice objects
+    """
+    return [Choice(name, value=value) for name, value in choices]
```

Comment:
This may be unnecessary abstraction but I found it annoying to always have to type that list comprehension out